src/utils.cpp

```diff
@@ -788,6 +788,16 @@ std::string exec_system(const char *cmd)
   return result;
 }
 
+std::string get_absolute_path(std::string const &path)
+{
+  int exist = ::access(path.c_str(), F_OK);
+  if (exist == -1)
+  {
+    return path;
+  }
+  return std_filesystem::canonical(path);
```

Comment:
TOCTOU here. How about using:

path canonical( const [std::filesystem::path](http://en.cppreference.com/w/cpp/filesystem/path)& p, [std::error_code](http://en.cppreference.com/w/cpp/error/error_code)& ec );

form of `canonical()` and explicitly checking for ENOENT?

Reply:
Fix it, please review, thanks.